File: backend/src/models/milk.rs
use chrono::NaiveDate;
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct MilkDayProduction {
    pub id: i32,
    pub animal_id: i32,
    pub date: NaiveDate,
    pub milk_amount: Option<f64>,
    pub avg_amount: Option<f64>,
    pub avg_weight: Option<f64>,
    pub isk: Option<f64>,
    pub created_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct MilkVisit {
    pub id: i32,
    pub animal_id: i32,
    pub visit_datetime: chrono::DateTime<chrono::Utc>,
    pub milk_amount: Option<f64>,
    pub duration_seconds: Option<i32>,
    pub milk_destination: Option<i32>,
    pub created_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct MilkQuality {
    pub id: i32,
    pub animal_id: i32,
    pub date: NaiveDate,
    pub milk_amount: Option<f64>,
    pub avg_amount: Option<f64>,
    pub avg_weight: Option<f64>,
    pub isk: Option<f64>,
    pub fat_percentage: Option<f64>,
    pub protein_percentage: Option<f64>,
    pub lactose_percentage: Option<f64>,
    pub scc: Option<i32>,
    pub milkings: Option<i32>,
    pub refusals: Option<i32>,
    pub failures: Option<i32>,
    pub created_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Deserialize)]
pub struct CreateMilkDayProduction {
    pub animal_id: i32,
    pub date: NaiveDate,
    pub milk_amount: Option<f64>,
    pub avg_amount: Option<f64>,
    pub avg_weight: Option<f64>,
    pub isk: Option<f64>,
}

impl CreateMilkDayProduction {
    pub fn validate(&self) -> Result<(), crate::errors::AppError> {
        use crate::validation::*;
        positive_i32(self.animal_id, "ID животного")?;
        date_not_future(&self.date, "Дата")?;
        opt_non_negative_f64(&self.milk_amount, "Надой")?;
        opt_non_negative_f64(&self.avg_amount, "Средний надой")?;
        opt_non_negative_f64(&self.avg_weight, "Средний вес")?;
        Ok(())
    }
}

#[derive(Debug, Deserialize)]
pub struct UpdateMilkDayProduction {
    pub date: Option<NaiveDate>,
    pub milk_amount: Option<f64>,
    pub avg_amount: Option<f64>,
    pub avg_weight: Option<f64>,
    pub isk: Option<f64>,
}

impl UpdateMilkDayProduction {
    pub fn validate(&self) -> Result<(), crate::errors::AppError> {
        use crate::validation::*;
        if let Some(ref d) = self.date {
            date_not_future(d, "Дата")?;
        }
        opt_non_negative_f64(&self.milk_amount, "Надой")?;
        opt_non_negative_f64(&self.avg_amount, "Средний надой")?;
        opt_non_negative_f64(&self.avg_weight, "Средний вес")?;
        Ok(())
    }
}

#[derive(Debug, Deserialize)]
pub struct MilkFilter {
    pub animal_id: Option<i32>,
    pub from_date: Option<NaiveDate>,
    pub till_date: Option<NaiveDate>,
    pub page: Option<i64>,
    pub per_page: Option<i64>,
}

File: frontend/src/lib/api/milk.ts
import { api, post, put, del, buildQuery } from './client';

export interface MilkDayProduction {
	id: number;
	animal_id: number;
	date: string;
	milk_amount: number | null;
	avg_amount: number | null;
	avg_weight: number | null;
	isk: number | null;
	created_at: string;
}

export interface MilkVisit {
	id: number;
	animal_id: number;
	visit_datetime: string;
	milk_amount: number | null;
	duration_seconds: number | null;
	milk_destination: number | null;
	created_at: string;
}

export interface MilkQuality {
	id: number;
	animal_id: number;
	date: string;
	milk_amount: number | null;
	avg_amount: number | null;
	avg_weight: number | null;
	isk: number | null;
	fat_percentage: number | null;
	protein_percentage: number | null;
	lactose_percentage: number | null;
	scc: number | null;
	milkings: number | null;
	refusals: number | null;
	failures: number | null;
	created_at: string;
}

export interface CreateMilkDayProduction {
	animal_id: number;
	date: string;
	milk_amount?: number;
	avg_amount?: number;
	avg_weight?: number;
	isk?: number;
}

export interface UpdateMilkDayProduction {
	date?: string;
	milk_amount?: number;
	avg_amount?: number;
	avg_weight?: number;
	isk?: number;
}

export interface MilkFilter {
	animal_id?: string;
	from_date?: string;
	till_date?: string;
	page?: number;
	per_page?: number;
}

export function listProductions(filter: MilkFilter = {}) {
	return api<{ data: MilkDayProduction[]; total: number }>(
		`/milk/day-productions${buildQuery(filter)}`,
	);
}

export function createProduction(data: CreateMilkDayProduction) {
	return post<{ data: MilkDayProduction }>('/milk/day-productions', data);
}

export function updateProduction(id: number, data: UpdateMilkDayProduction) {
	return put<{ data: MilkDayProduction }>(`/milk/day-productions/${id}`, data);
}

export function deleteProduction(id: number) {
	return del<{ message: string }>(`/milk/day-productions/${id}`);
}

export function listVisits(filter: MilkFilter = {}) {
	return api<{ data: MilkVisit[]; total: number }>(`/milk/visits${buildQuery(filter)}`);
}

export function listQuality(filter: MilkFilter = {}) {
	return api<{ data: MilkQuality[]; total: number }>(`/milk/quality${buildQuery(filter)}`);
}
